File: solutions/csharp/0872-leaf-similar-tree.cs
public class Solution {
    public bool LeafSimilar(TreeNode root1, TreeNode root2) {
        IList<int> leaves1 = new List<int>();
        IList<int> leaves2 = new List<int>();
        Dfs(root1, leaves1);
        Dfs(root2, leaves2);
        return leaves1.SequenceEqual(leaves2);
    }

    public void Dfs(TreeNode node, IList<int> leafValues) {
        if (node != null) {
            if (node.left == null && node.right == null) {
                leafValues.Add(node.val);
            }
            Dfs(node.left, leafValues);
            Dfs(node.right, leafValues);
        }
    }
}

File: solutions/python/0872-leaf-similar-tree.py
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def dfs(node): 
            if node: 
                if not node.left and not node.right: 
                    yield node.val 
                yield from dfs(node.left) 
                yield from dfs(node.right) 

        return list(dfs(root1)) == list(dfs(root2)) 
